x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java

```diff
@@ -334,33 +332,36 @@ private MethodSpec addRawInputExploded(boolean hasMask) {
         }
 
         for (int i = 0; i < aggParams.size(); i++) {
-            AggregationParameter p = aggParams.get(i);
-            builder.addStatement("$T $L = page.getBlock(channels.get($L))", blockType(p.type()), p.blockName(), i);
+            Argument a = aggParams.get(i);
+            builder.addStatement("$T $L = page.getBlock(channels.get($L))", a.dataType(true), a.blockName(), i);
         }
-        for (AggregationParameter p : aggParams) {
-            builder.addStatement("$T $L = $L.asVector()", vectorType(p.type()), p.vectorName(), p.blockName());
-            builder.beginControlFlow("if ($L == null)", p.vectorName());
+
+        boolean isBlockArgument = aggParams.getFirst() instanceof BlockArgument;
+
+        for (Argument a : aggParams) {
+            String rawBlock = "addRawBlock("
+                + aggParams.stream().map(arg -> arg.blockName()).collect(joining(", "))
+                + (hasMask ? ", mask" : "")
+                + ")";
+
+            if (isBlockArgument == false) {
+                a.resolveVectors(builder, rawBlock, "return");
+            } else {
+                builder.addStatement(rawBlock);
+            }
+        }
+
+        if (isBlockArgument == false) {
             builder.addStatement(
-                "addRawBlock("
-                    + aggParams.stream().map(AggregationParameter::blockName).collect(joining(", "))
-                    + (hasMask ? ", mask" : "")
-                    + ")"
+                "addRawVector(" + aggParams.stream().map(a -> a.vectorName()).collect(joining(", ")) + (hasMask ? ", mask" : "") + ")"
```

Comment:
For this I think you want "if *all* arguments are BlockArgument` then call `addRawBlock` - otherwise you keep the `addRawVector`. I don't think it changes the generated code any, but feels a little more clear around the noop above.